### ISSUE TITLE
Feature Request - Subsuming DOM

### ISSUE THREAD

**Issue by kitsonk** [AFASSoftware/maquette#34]
@pottedmeat and myself were having a discussion last night and we have identified a use case that we were wondering what the Maquette team's opinion was.

As we are all aware, quick application startup, especially on mobile, is important.  So lets say we were able to calculate the state of the DOM of a complex web app server side (e.g. using Maquette with JSDom) and then were to send that to the client as our initial page.  We would then load maquette and replicate the Projector and then load the rest of our components, set their state and we would then be ready to render client side.

We could of course do that now, but my understanding is that the Projector would only be able to detach the existing node and replace it with one it calculated and then from that point you would get the efficiency of DOM diffing on subsequent applications.

When in reality, the pre calculated initial state of the DOM I sent from the server would actually be the initial calculated structure (minus maybe some event handlers).

I was wondering what the thoughts are about enabling a feature (maybe even as a seperate module) that creates a Projector that can subsume a set of DOM, essentially matching the VNodes up with existing DOM and then projecting from there.

**Comment by johan-gorter** [comment]
It is a good point that you raise here, and it is definitely something that we want to investigate further in the future. Let me summarize the options that we are considering:

1) No server-side rendering
First of all, maquette was built to be fast and lightweight. Compared to React, a mobile application should therefore start up way faster than with React. It is however unlikely that it would outperform server-side rendering.

2) Server-side html rendering, without adopting/subsuming DOM nodes
If you generate the DOM server-side using JSDom (or a specialized VNode to html function), you can use the existing Projector.replace() function to replace the DOM. I do not think that subsuming/adopting the server-rendered DOM nodes would be beneficial for the performance. I recommend that the server renders all controls and buttons disabled/readonly. Links could work though.

3) Server-side html rendering with adopting/subsuming DOM nodes
Reusing the server-rendered DOM nodes looks tempting. You can keep the focus/cursor and the user can start entering texts and toggle switches before the application is fully loaded. This is however very complex to implement and the user may get confused if toggling switches does not immediately have effect.

4) Server-side hyperscript rendering
We could also render HTML with an inline-script containing a small portion of maquette code followed by a piece of server-side rendered Hyperscript. The Hyperscript is by nature more compact than HTML, which could compensate the overhead of a small portion of maquette code. This approach makes adopting/subsuming the VNodes easier. The remaining problems here would be:
- The key property on the VNodes must be a string or a number to be able to serialize it
- Event handlers will be added after fully loading. A generic event-buffering event-handler could be used for the the server-side rendered DOM, but disabling the controls may still lead to a better user experience. But if we disable the controls, then subsuming/adopting VNodes may not have any benefits.

My conclusions thusfar:

- Having the server render HTML with disabled controls and replacing the DOM completely is the easiest to implement and this is probably enough to give the user a good user-experience. This may also yield the best SEO results.
 - Subsuming the DOM is hard to implement and I doubt that it will have performance benefits.
 - When rendering non-readonly controls subsuming the DOM will be even harder to implement

It looks like you will be needing this functionality before we need it at @AFASSoftware. I hope you will be sharing your endeavours with us.

**Comment by kitsonk** [comment]
@johan-gorter a very early work in progress prototype is available at [kitsonk/widgets](https://github.com/kitsonk/widgets).  In particular [src/util/vdom.ts](https://github.com/kitsonk/widgets/blob/master/src/util/vdom.ts) is the "abstraction" for Maquette that manages the relationship between the widgets/components and the projectors.  For context, it is part of the direction of the [Dojo 2 Application Architecture](https://docs.google.com/document/d/1Kgc99f8yTtZGopKr6OWGMjsJTV5hHNlr0ztWQS_cRnA/edit#)

As far as the conclusions, I think I agree that potentially the most effective would be to:
* Render server side an HTML document.  Stated but disabled would give the visual indication to the user that the application is available, but not interactive (as you mention).
* Deliver that to the client and instantiate the widgets with state (this is the most costly activity).
* Pass to a projector that replaces the document.  As you say, I one of redraw after load is unlikely to be too much of a burden and it won't occur until the application is actually ready to interact with the user.

I just looked and I did notice that a replace seems to be a efficient as possible, in that it does create the entire DOM structure before removing the child.  So I think you are right, that should be more than acceptable from a user experience.
